src/ZfcRbac/Factory/ObjectRepositoryRoleProviderFactory.php

```diff
@@ -40,42 +39,55 @@ class ObjectRepositoryRoleProviderFactory implements FactoryInterface, MutableCr
     /**
      * {@inheritDoc}
      */
-    public function setCreationOptions(array $options)
+    public function __construct(array $options = [])
     {
         $this->options = $options;
     }
 
     /**
-     * {@inheritDoc}
+     * @param ContainerInterface $container
      * @return ObjectRepositoryRoleProvider
      */
-    public function createService(ServiceLocatorInterface $serviceLocator)
+    public function __invoke(ContainerInterface $container, $resolvedName, $options)
     {
-        $parentLocator    = $serviceLocator->getServiceLocator();
+        if (! method_exists($container, 'build')) { // servicemanager v3
+            $container = $container->getServiceLocator();
+            $options = $this->options;
+        }
+
         $objectRepository = null;
 
-        if (!isset($this->options['role_name_property'])) {
+        if (!isset($options['role_name_property'])) {
             throw new Exception\RuntimeException('The "role_name_property" option is missing');
         }
 
-        if (isset($this->options['object_repository'])) {
+        if (isset($options['object_repository'])) {
             /* @var \Doctrine\Common\Persistence\ObjectRepository $objectRepository */
-            $objectRepository = $parentLocator->get($this->options['object_repository']);
+            $objectRepository = $container->get($options['object_repository']);
 
-            return new ObjectRepositoryRoleProvider($objectRepository, $this->options['role_name_property']);
+            return new ObjectRepositoryRoleProvider($objectRepository, $options['role_name_property']);
         }
 
-        if (isset($this->options['object_manager']) && isset($this->options['class_name'])) {
+        if (isset($options['object_manager']) && isset($options['class_name'])) {
             /* @var \Doctrine\Common\Persistence\ObjectManager $objectManager */
-            $objectManager    = $parentLocator->get($this->options['object_manager']);
-            $objectRepository = $objectManager->getRepository($this->options['class_name']);
+            $objectManager    = $container->get($options['object_manager']);
+            $objectRepository = $objectManager->getRepository($options['class_name']);
 
-            return new ObjectRepositoryRoleProvider($objectRepository, $this->options['role_name_property']);
+            return new ObjectRepositoryRoleProvider($objectRepository, $options['role_name_property']);
         }
 
         throw new Exception\RuntimeException(
             'No object repository was found while creating the ZfcRbac object repository role provider. Are
              you sure you specified either the "object_repository" option or "object_manager"/"class_name" options?'
         );
     }
+
+    /**
+     * {@inheritDoc}
+     * @return ObjectRepositoryRoleProvider
+     */
+    public function createService(ServiceLocatorInterface $serviceLocator)
+    {
+        return $this($serviceLocator, ObjectRepositoryRoleProvider::class);
```

Comment:
I would suggest keep `__invoke()` clean only what required for SMv3 because this will be called only with SMv3 and do all required changes here. For example:
```php
return $this($serviceLocator->getServiceLocator(), ObjectRepositoryRoleProvider::class, $this->options);
```